Sources/GRPC/Interceptor/ClientInterceptorContext.swift

```diff
@@ -0,0 +1,145 @@
+  private var nextInbound: ClientInterceptorContext<Request, Response>? {
+    return self.pipeline.context(atIndex: self.index + 1)
+  }
+
+  // The next context in the outbound direction, if one exists.
+  private var nextOutbound: ClientInterceptorContext<Request, Response>? {
+    return self.pipeline.context(atIndex: self.index - 1)
+  }
+
+  /// The `EventLoop` this interceptor pipeline is being executed on.
+  public var eventLoop: EventLoop {
+    return self.pipeline.eventLoop
+  }
+
+  /// A logger.
+  public var logger: Logger {
+    return self.pipeline.logger
+  }
+
+  /// Construct a `ClientInterceptorContext` for the interceptor at the given index within in
+  /// interceptor pipeline.
+  internal init(pipeline: ClientInterceptorPipeline<Request, Response>, index: Int) {
+    self.pipeline = pipeline
+    self.index = index
+  }
+
+  /// Forwards the response part to the next inbound interceptor in the pipeline, if there is one.
+  ///
+  /// - Parameter part: The response part to forward.
+  /// - Note: This does *not* have to be called from the `EventLoop`.
+  public func read(_ part: ClientResponsePart<Response>) {
+    if self.eventLoop.inEventLoop {
+      self._read(part)
+    } else {
+      self.eventLoop.execute {
+        self._read(part)
+      }
+    }
+  }
+
+  /// Forwards the request part to the next outbound interceptor in the pipeline, if there is one.
+  ///
+  /// - Parameters:
+  ///   - part: The request part to forward.
+  ///   - promise: The promise the complete when the part has been written.
+  /// - Note: This does *not* have to be called from the `EventLoop`.
+  public func write(
+    _ part: ClientRequestPart<Request>,
+    promise: EventLoopPromise<Void>?
+  ) {
+    if self.eventLoop.inEventLoop {
+      self._write(part, promise: promise)
+    } else {
+      self.eventLoop.execute {
+        self._write(part, promise: promise)
+      }
+    }
+  }
+
+  /// Forwards a request to cancel the RPC to the next outbound interceptor in the pipeline.
+  ///
+  /// - Parameter promise: The promise to complete with the outcome of the cancellation request.
+  /// - Note: This does *not* have to be called from the `EventLoop`.
+  public func cancel(promise: EventLoopPromise<Void>?) {
+    if self.eventLoop.inEventLoop {
+      self._cancel(promise: promise)
+    } else {
+      self.eventLoop.execute {
+        self._cancel(promise: promise)
+      }
+    }
+  }
+}
+
+extension ClientInterceptorContext {
+  private func _read(_ part: ClientResponsePart<Response>) {
+    self.eventLoop.assertInEventLoop()
+    self.nextInbound?.invokeRead(part)
+  }
+
+  private func _write(
+    _ part: ClientRequestPart<Request>,
+    promise: EventLoopPromise<Void>?
+  ) {
+    self.eventLoop.assertInEventLoop()
+
+    if let outbound = self.nextOutbound {
+      outbound.invokeWrite(part, promise: promise)
+    } else {
+      promise?.fail(GRPCStatus(code: .unavailable, message: "The RPC has already completed"))
```

Comment:
Is this right? There seems to be some inconsistency between `_read` and `_write` that I don't fully understand about when the next interceptor might be `nil`.

Reply:
So `_read` is more like NIOs `fireChannelRead` rather than NIOs `read`.

What's not implemented here is removing the contexts when the RPC completes, that's when we expect them to be `nil`.